processing/src/main/java/org/apache/druid/segment/QueryableIndexCursorSequenceBuilder.java

```diff
@@ -0,0 +1,638 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.segment;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Function;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import org.apache.druid.collections.bitmap.ImmutableBitmap;
+import org.apache.druid.java.util.common.granularity.Granularity;
+import org.apache.druid.java.util.common.guava.Sequence;
+import org.apache.druid.java.util.common.guava.Sequences;
+import org.apache.druid.java.util.common.io.Closer;
+import org.apache.druid.query.BaseQuery;
+import org.apache.druid.query.filter.Filter;
+import org.apache.druid.query.monomorphicprocessing.RuntimeShapeInspector;
+import org.apache.druid.segment.column.BaseColumn;
+import org.apache.druid.segment.column.ColumnHolder;
+import org.apache.druid.segment.column.NumericColumn;
+import org.apache.druid.segment.data.Offset;
+import org.apache.druid.segment.data.ReadableOffset;
+import org.apache.druid.segment.historical.HistoricalCursor;
+import org.apache.druid.segment.vector.BitmapVectorOffset;
+import org.apache.druid.segment.vector.FilteredVectorOffset;
+import org.apache.druid.segment.vector.NoFilterVectorOffset;
+import org.apache.druid.segment.vector.QueryableIndexVectorColumnSelectorFactory;
+import org.apache.druid.segment.vector.VectorColumnSelectorFactory;
+import org.apache.druid.segment.vector.VectorCursor;
+import org.apache.druid.segment.vector.VectorOffset;
+import org.joda.time.DateTime;
+import org.joda.time.Interval;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+public class QueryableIndexCursorSequenceBuilder
+{
+  /**
+   * At this threshold, timestamp searches switch from binary to linear. See
+   * {@link #timeSearch(NumericColumn, long, int, int, int)} for more details.
+   */
+  private static final int TOO_CLOSE_FOR_MISSILES = 15000;
```

Comment:
[![jester's dead](https://img.youtube.com/vi/JPgMyh7ei6w/0.jpg)](https://www.youtube.com/watch?v=JPgMyh7ei6w)

Reply:
Just wonder how that value was chosen?

Reply:
See this comment from the `timeSearch` method:

> The idea is to avoid too much decompression buffer thrashing. The default value `TOO_CLOSE_FOR_MISSILES` is chosen to be similar to the typical number of timestamps per block.

I moved the sentence about choice of default value to the javadoc for `TOO_CLOSE_FOR_MISSILES`, and kept the "idea" comment in the javadoc for `timeSearch`.

Reply:
It's not clear why `the typical number of timestamps per block` is 15000?

Reply:
> It's not clear why the typical number of timestamps per block is 15000?

Each block in a long typed column is 64KB, so if a timestamp can be stored in 3-4 bytes (typical with delta encoding) there's roughly that number per block. Or really a bit more.

But actually, you know what, I had never benchmarked it until today. I just benchmarked it and it looks like the pure binary search is fastest anyway. So, I removed the switching and made the search purely binary.

Benchmark results:

```
Benchmark                (query)  (rowsPerSegment)    (timestampString)  (tooCloseForMissiles)  (vectorize)  Mode  Cnt   Score    Error  Units
timeSearch                    15           5000000                 2000                      0        force  avgt   10   0.647 ±  0.001  ms/op
timeSearch                    15           5000000                 2000                   5000        force  avgt   10   0.643 ±  0.001  ms/op
timeSearch                    15           5000000                 2000                  15000        force  avgt   10   0.643 ±  0.001  ms/op
timeSearch                    15           5000000                 2000                  25000        force  avgt   10   0.579 ±  0.001  ms/op
timeSearch                    15           5000000                 2000                  50000        force  avgt   10   0.514 ±  0.001  ms/op
timeSearch                    15           5000000                 2000               99999999        force  avgt   10  ≈ 10⁻⁴           ms/op
timeSearch                    15           5000000  2000-01-01T01:00:00                      0        force  avgt   10   0.578 ±  0.001  ms/op
timeSearch                    15           5000000  2000-01-01T01:00:00                   5000        force  avgt   10   0.589 ±  0.001  ms/op
timeSearch                    15           5000000  2000-01-01T01:00:00                  15000        force  avgt   10   0.588 ±  0.001  ms/op
timeSearch                    15           5000000  2000-01-01T01:00:00                  25000        force  avgt   10   0.751 ±  0.008  ms/op
timeSearch                    15           5000000  2000-01-01T01:00:00                  50000        force  avgt   10   0.607 ±  0.004  ms/op
timeSearch                    15           5000000  2000-01-01T01:00:00               99999999        force  avgt   10   2.106 ±  0.078  ms/op
timeSearch                    15           5000000     2000-01-01T10:30                      0        force  avgt   10   0.708 ±  0.001  ms/op
timeSearch                    15           5000000     2000-01-01T10:30                   5000        force  avgt   10   0.722 ±  0.002  ms/op
timeSearch                    15           5000000     2000-01-01T10:30                  15000        force  avgt   10   0.728 ±  0.001  ms/op
timeSearch                    15           5000000     2000-01-01T10:30                  25000        force  avgt   10   0.773 ±  0.016  ms/op
timeSearch                    15           5000000     2000-01-01T10:30                  50000        force  avgt   10   0.854 ±  0.001  ms/op
timeSearch                    15           5000000     2000-01-01T10:30               99999999        force  avgt   10  21.863 ±  0.146  ms/op
timeSearch                    15           5000000     2000-01-01T16:30                      0        force  avgt   10   0.643 ±  0.001  ms/op
timeSearch                    15           5000000     2000-01-01T16:30                   5000        force  avgt   10   0.653 ±  0.001  ms/op
timeSearch                    15           5000000     2000-01-01T16:30                  15000        force  avgt   10   0.668 ±  0.001  ms/op
timeSearch                    15           5000000     2000-01-01T16:30                  25000        force  avgt   10   0.685 ±  0.004  ms/op
timeSearch                    15           5000000     2000-01-01T16:30                  50000        force  avgt   10   0.935 ±  0.013  ms/op
timeSearch                    15           5000000     2000-01-01T16:30               99999999        force  avgt   10  39.498 ±  0.022  ms/op
timeSearch                    15           5000000           2000-01-02                      0        force  avgt   10   0.602 ±  0.001  ms/op
timeSearch                    15           5000000           2000-01-02                   5000        force  avgt   10   0.616 ±  0.003  ms/op
timeSearch                    15           5000000           2000-01-02                  15000        force  avgt   10   0.645 ±  0.001  ms/op
timeSearch                    15           5000000           2000-01-02                  25000        force  avgt   10   0.707 ±  0.001  ms/op
timeSearch                    15           5000000           2000-01-02                  50000        force  avgt   10   0.811 ±  0.001  ms/op
timeSearch                    15           5000000           2000-01-02               99999999        force  avgt   10  61.194 ±  0.036  ms/op
```

Reply:
@egor-ryashin What do you think of the latest changes?

Reply:
Removing **tooCloseForMissiles** check, isn't it an equivalent of setting `tooCloseForMissiles = 99999999 ` 
 which shows slower operations?

Reply:
@egor-ryashin It would be the equivalent of tooCloseForMissiles = 0, since I removed a check that did a “break” out of the binary search when maxIndex and minIndex were less than tooCloseForMissiles apart.

 (the threshold was: when search is this close, switch to linear. So 99999999 means fully linear and 0 means fully binary)